### PR TITLE
[Feature] Support SoftTeacher

### PR DESCRIPTION
**Pull request by Czm369**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: configs/semi/faster_rcnn/semi_base_faster_rcnn_r50_caffe_fpn_180k_partial_coco.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+_base_ = [
+    '../../_base_/models/faster_rcnn_r50_fpn.py',
+    '../../_base_/default_runtime.py', '../semi_partial_coco_detection.py'
+]
+
+model_wrapper = dict(
```
Comment (RangiLyu) [inline configs/semi/faster_rcnn/semi_base_faster_rcnn_r50_caffe_fpn_180k_partial_coco.py:6]: `model_wrapper` should be distributed model wrapper such as MMDDP, not a detector.

Comment (RangiLyu) [inline configs/semi/faster_rcnn/semi_base_faster_rcnn_r50_caffe_fpn_180k_partial_coco.py:6]: model_wrapper? Where will it be built?

**Review by RangiLyu — COMMENTED** [review]
How will we organize the folder?

File: configs/semi/faster_rcnn/semi_base_faster_rcnn_r50_caffe_fpn_180k_partial_coco.py (hunk 1 above)
Comment (Czm369) [reply configs/semi/faster_rcnn/semi_base_faster_rcnn_r50_caffe_fpn_180k_partial_coco.py:6]: `model_wrapper` will be replaced by `model` in https://github.com/open-mmlab/mmdetection/blob/3b72b12fe9b14de906d1363982b9fba05e7d47c1/mmdet/utils/replace_cfg_vals.py#L67

File: mmdet/models/detectors/semi_base.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        multi_batch_data_samples[
+            'unsup_student'] = self.project_pseudo_instances(
+                origin_pseudo_instances,
+                multi_batch_data_samples['unsup_student'])
+        losses.update(**self.pseudo_loss(
+            multi_batch_inputs['unsup_student'],
+            multi_batch_data_samples['unsup_student'], batch_info))
+        return losses
+
+    def gt_loss(self, batch_inputs: Tensor,
```
Comment (hhaAndroid) [inline mmdet/models/detectors/semi_base.py:93]: supervised_loss

File: mmdet/models/detectors/semi_base.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        """
+        gt_loss = {
+            'sup_' + k: v
+            for k, v in self.weight(
+                self.student.loss(batch_inputs, batch_data_samples),
+                self.semi_train_cfg.get('sup_weight', 1.)).items()
+        }
+        return gt_loss
+
+    def pseudo_loss(self,
```
Comment (hhaAndroid) [inline mmdet/models/detectors/semi_base.py:116]: pseudo_supervised_loss

File: mmdet/models/detectors/semi_base.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+        pseudo_loss = {
+            'unsup_' + k: v
+            for k, v in self.weight(
+                self.student.loss(batch_inputs, batch_data_samples),
+                unsup_weight).items()
+        }
+        return pseudo_loss
+
+    @staticmethod
+    def weight(losses: dict, weight: float) -> dict:
```
Comment (hhaAndroid) [inline mmdet/models/detectors/semi_base.py:158]: weight ->branch_weights

File: mmdet/models/detectors/semi_base.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    @torch.no_grad()
+    def get_pseudo_instances(
+            self, batch_inputs: Tensor, batch_data_samples: SampleList
+    ) -> Tuple[SampleList, Optional[dict]]:
+        """Get pseudo instances from teacher model."""
+        results_list = self.teacher(
+            batch_inputs, batch_data_samples, mode='predict')
+        for data_samples, results in zip(batch_data_samples, results_list):
+            data_samples.gt_instances = results.pred_instances
+        batch_info = None
```
Comment (hhaAndroid) [inline mmdet/models/detectors/semi_base.py:190]: is need?

File: mmdet/models/detectors/semi_base.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+                or `pseudo_sem_seg` in fact.
+            batch_info (dict): Batch information of teacher model
+                forward propagation process. Defaults to None.
+
+        Returns:
+            dict: A dictionary of loss components
+        """
+        for data_samples in batch_data_samples:
+            if data_samples.gt_instances.bboxes.shape[0] > 0:
+                data_samples.gt_instances = data_samples.gt_instances[
```
Comment (hhaAndroid) [inline mmdet/models/detectors/semi_base.py:138]: Why not put it in `filter_pseudo_instances`？

File: configs/semi/faster_rcnn/semi_base_faster_rcnn_r50_caffe_fpn_180k_partial_coco.py (hunk 1 above)
Comment (RangiLyu) [reply configs/semi/faster_rcnn/semi_base_faster_rcnn_r50_caffe_fpn_180k_partial_coco.py:6]: mmengine has a similar cfg called `model_wrapper_cfg` which is for DDP wrappers.

Comment (RangiLyu) [reply configs/semi/faster_rcnn/semi_base_faster_rcnn_r50_caffe_fpn_180k_partial_coco.py:6]: Need to add some comments about this logic in the config.

File: mmdet/models/detectors/semi_base.py (hunk 3 above)
Comment (Czm369) [reply mmdet/models/detectors/semi_base.py:116]: loss_by_pseudo_instance ?

File: mmdet/models/detectors/semi_base.py (hunk 5 above)
Comment (Czm369) [reply mmdet/models/detectors/semi_base.py:190]: Subclasses need this variable

File: mmdet/models/detectors/semi_base.py (hunk 6 above)
Comment (Czm369) [reply mmdet/models/detectors/semi_base.py:138]: Subclasses need filtering pseudo instances by different scores.

Comment (Czm369) [reply mmdet/models/detectors/semi_base.py:138]: But all semi-model need filter too small or invalid pseudo instances.

File: mmdet/models/detectors/semi_base.py (hunk 3 above)
Comment (ZwwWayne) [reply mmdet/models/detectors/semi_base.py:116]: loss_by_pseudo_label might be more clear

File: configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    style='caffe',
+    init_cfg=dict(
+        type='Pretrained',
+        checkpoint='open-mmlab://detectron2/resnet50_caffe'))
+
+model = dict(
+    _delete_=True,
+    type='SemiBaseDetector',
+    detector=detector,
+    semi_train_cfg=dict(
```
Comment (hhaAndroid) [inline configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py:30]: Can't it be called train_cfg?

File: configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        freeze_teacher=True,
+        sup_weight=1.0,
+        unsup_weight=4.0,
+        cls_pseudo_thr=0.9,
+        min_pseudo_bbox_wh=(1e-2, 1e-2)),
+    semi_test_cfg=dict(predict_on='teacher'),
+    data_preprocessor=dict(
```
Comment (hhaAndroid) [inline configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py:37]: Better to put it in the front

File: configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    _delete_=True,
+    type='SemiBaseDetector',
+    detector=detector,
+    semi_train_cfg=dict(
+        freeze_teacher=True,
+        sup_weight=1.0,
+        unsup_weight=4.0,
+        cls_pseudo_thr=0.9,
+        min_pseudo_bbox_wh=(1e-2, 1e-2)),
+    semi_test_cfg=dict(predict_on='teacher'),
```
Comment (hhaAndroid) [inline configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py:36]: Add a note to explain the meaning

File: configs/soft_teacher/semi_partial_coco_detection.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        type='PackDetInputs',
+        meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
+                   'scale_factor'))
+]
+
+fold = 1
+percent = 10
+
+train_dataloader = dict(
+    batch_size=5,
```
Comment (hhaAndroid) [inline configs/soft_teacher/semi_partial_coco_detection.py:86]: If batch_size is related to source_ratio, it is best to write it out separately

File: configs/soft_teacher/semi_partial_coco_detection.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
+# dataset settings
```
Comment (ZwwWayne) [inline configs/soft_teacher/semi_partial_coco_detection.py:1]: suggest move semi_partial_coco_detection to _base_/datasets/

File: mmdet/models/detectors/semi_base.py (hunk 4 above)
Comment (ZwwWayne) [reply mmdet/models/detectors/semi_base.py:155]: use reweight_loss

File: mmdet/models/detectors/soft_teacher.py [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+                torch.zeros_like(data_sample.gt_instances.labels)
+
+        rpn_losses, rpn_results_list = self.student.rpn_head.loss_and_predict(
+            x, rpn_data_samples, proposal_cfg=proposal_cfg)
+        for key in rpn_losses.keys():
+            if 'loss' in key and 'rpn' not in key:
+                rpn_losses[f'rpn_{key}'] = rpn_losses.pop(key)
+        return rpn_losses, rpn_results_list
+
+    def unsup_rcnn_cls_loss(self, x: Tuple[Tensor],
```
Comment (ZwwWayne) [inline mmdet/models/detectors/soft_teacher.py:175]: should we also use loss_rcnn_cls_by_pseudo_labels?

Comment (Czm369) [reply mmdet/models/detectors/soft_teacher.py:175]: rcnn_cls_loss_by_pseudo_instances ?

File: configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py (hunk 7 above)
Comment (Czm369) [reply configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py:30]: It aims to diff from `train_cfg` of `Faster RCNN` to avoid confusion confusion

File: configs/_base_/datasets/semi_partial_coco_detection.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
+# dataset settings
```
Comment (ZwwWayne) [inline configs/_base_/datasets/semi_partial_coco_detection.py:1]: what does the name partial stand for?

File: configs/_base_/datasets/semi_partial_coco_detection.py [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+    dict(type='LoadImageFromFile', file_client_args=file_client_args),
+    dict(type='LoadAnnotations', with_bbox=True),
+    dict(type='RandomResize', scale=scale, keep_ratio=True),
+    dict(type='RandomFlip', prob=0.5),
+    dict(type='RandAugment', aug_space=color_space, aug_num=1),
+    dict(type='FilterAnnotations', min_gt_bbox_wh=(1e-2, 1e-2)),
+    dict(type='MultiBranch', sup=dict(type='PackDetInputs'))
+]
+
+weak_pipeline = [
```
Comment (ZwwWayne) [inline configs/_base_/datasets/semi_partial_coco_detection.py:42]: since there are too many pipelines, we should add comments above them to indicate their usages. For example, pipeline used to augment xxx, which will be sent to teacher/student

File: configs/_base_/datasets/semi_partial_coco_detection.py [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+        type='PackDetInputs',
+        meta_keys=('img_id', 'img_path', 'ori_shape', 'img_shape',
+                   'scale_factor'))
+]
+
+batch_size = 5
+num_workers = 5
+# The format of labeled_ann_file and unlabeled_ann_file are
+# instances_train2017.{fold}@{percent}.json, and
+# instances_train2017.{fold}@{percent}-unlabeled.json
```
Comment (ZwwWayne) [inline configs/_base_/datasets/semi_partial_coco_detection.py:93]: do we really need unlabled json? Can we simply use an option to avoid loading annotations.

File: configs/_base_/datasets/semi_partial_coco_detection.py (hunk 13 above)
Comment (Czm369) [reply configs/_base_/datasets/semi_partial_coco_detection.py:1]: Part of coco train2017 will be divided into labeled data, and the remaining data will be used as unlabeled data.

File: configs/_base_/datasets/semi_partial_coco_detection.py (hunk 15 above)
Comment (Czm369) [reply configs/_base_/datasets/semi_partial_coco_detection.py:93]: We need at least one annotation of unlabeled data, so that the dataset can get the path of each image, and the total number of images.

File: configs/_base_/datasets/semi_partial_coco_detection.py (hunk 13 above)
Comment (ZwwWayne) [reply configs/_base_/datasets/semi_partial_coco_detection.py:1]: than partial can be removed, semi_coco_detection is clear enough to indicate that the config is used for semi-supervised setting, essentially the whole dataset is used except that only part of labels is used.

File: configs/soft_teacher/soft_teacher_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py [hunk 16]
```diff
@@ -0,0 +1,11 @@
+_base_ = ['semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py']
```
Comment (ZwwWayne) [inline configs/soft_teacher/soft_teacher_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py:1]: soft_teacher_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py -> faster-rcnn_r50-caffe-fpn_softteacher-180k_coco.py

File: configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py [hunk 17]
```diff
@@ -0,0 +1,69 @@
+_base_ = [
```
Comment (ZwwWayne) [inline configs/soft_teacher/semi_base_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py:1]: Merge this config with the config of softteacher

File: configs/_base_/datasets/semi_partial_coco_detection.py [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+# Default fold and percent are 1 and 10, which means 10% data
+# of coco train2017 will be selected as labeled data,
+# and the remaining data will be used as unlabeled data.
+labeled_ann_file = 'coco_semi_anns/instances_train2017.1@10.json'
```
Comment (ZwwWayne) [inline configs/_base_/datasets/semi_partial_coco_detection.py:103]: coco_semi_anns -> coco/semi_anns/

File: configs/_base_/datasets/semi_partial_coco_detection.py [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+    sampler=dict(
+        type='GroupMultiSourceSampler',
+        batch_size=batch_size,
+        source_ratio=[1, 4]),
+    dataset=dict(
+        type='ConcatDataset',
+        datasets=[
+            dict(
+                type=dataset_type,
+                data_root='data/',
```
Comment (ZwwWayne) [inline configs/_base_/datasets/semi_partial_coco_detection.py:119]: should use the same data_root 'data/coco'

File: configs/_base_/datasets/semi_partial_coco_detection.py (hunk 13 above)
Comment (ZwwWayne) [reply configs/_base_/datasets/semi_partial_coco_detection.py:1]: may indicate 1@10 in the file name, use semi-0.1_coco_detection.py

File: tools/misc/split_coco.py [hunk 20]
```diff
@@ -23,7 +23,7 @@ def parse_args():
         '--out-dir',
         type=str,
         help='The output directory of coco semi-supervised annotations.',
-        default='./data/coco_semi_annos/')
+        default='./data/coco_semi_anns/')
```
Comment (ZwwWayne) [inline tools/misc/split_coco.py:26]: data/coco/semi_anns

File: configs/soft_teacher/soft_teacher_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py (hunk 16 above)
Comment (ZwwWayne) [reply configs/soft_teacher/soft_teacher_faster-rcnn_r50_caffe_fpn_180k_partial_coco.py:1]: or consider soft-teacher_faster-rcnn_r50-caffe-fpn_180k_semi-0.1-coco.py

**Review by ZwwWayne — APPROVED** [review]
